qiskit_machine_learning/kernels/quantum_kernel.py

```diff
@@ -0,0 +1,322 @@
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Quantum Kernel Algorithm"""
+
+from typing import Optional, Union
+
+import logging
+
+import numpy as np
+
+from qiskit import QuantumCircuit, QuantumRegister, ClassicalRegister
+from qiskit.circuit import ParameterVector
+from qiskit.providers import Backend, BaseBackend
+
+from qiskit.utils import QuantumInstance
+from qiskit.algorithms import AlgorithmError
+
+logger = logging.getLogger(__name__)
+
+
+class QuantumKernel:
+    r"""Quantum Kernel.
+
+    The general task of machine learning is to find and study patterns in data. For many
+    algorithms, the datapoints are better understood in a higher dimensional feature space,
+    through the use of a kernel function:
+
+    .. math::
+        K(x, y) = \langle f(x), f(y)\rangle.
+
+    Here K is the kernel function, x, y are n dimensional inputs. f is a map from n-dimension
+    to m-dimension space. :math:`\langle x, y \rangle` denotes the dot product.
+    Usually m is much larger than n.
+
+    The quantum kernel algorithm calculates a kernel matrix, given datapoints x and y and feature
+    map f, all of n dimension. This kernel matrix can then be used in classical machine learning
+    algorithms such as support vector classification, spectral clustering or ridge regression.
+    """
+
+    def __init__(self,
+                 feature_map: QuantumCircuit,
+                 enforce_psd: bool = False,
+                 batch_size: int = 1000,
+                 quantum_instance: Optional[
+                     Union[QuantumInstance, BaseBackend, Backend]] = None) -> None:
+        """
+        Args
+            feature_map:       # parameterized circuit to be used as the feature map
+            enforce_psd:       # project to closest positive semidefinite matrix if x = y
+            batch_size:        # number of circuits to batch together for computation
+            quantum_instance:  # Quantum Instance or Backend
+        """
+        self._feature_map = feature_map
+        self._enforce_psd = enforce_psd
+        self._batch_size = batch_size
+        self._quantum_instance = quantum_instance
+
+    @property
+    def feature_map(self) -> QuantumCircuit:
+        """ Returns feature map """
+        return self._feature_map
+
+    @feature_map.setter
+    def feature_map(self, feature_map: QuantumCircuit):
+        """ Sets feature map """
+        self._feature_map = feature_map
+        if not hasattr(feature_map, 'ordered_parameters'):
+            self.feature_map.ordered_parameters = sorted(feature_map.parameters,
```

Comment:
Parameters now have a fixed order in Terra's parameterized circuits such that I think we can drop the sorted here and rely on the sorting in parameterized circuits now - which is better. @Cryoris can correct me. Also the ordered_parameters field, as far as I am aware, has now been deprecated,